Live metrics docs (#4012)
iterative/dvc.org: content/docs/studio/user-guide/projects-and-experiments/live-metrics.md

# Generate live (real-time) metrics for experiments

In your model training CI action, you can use [DVCLive] to send live updates to
metrics and plots back to Iterative Studio, without writing them to your Git
repository. This requires a 3-step process:

1. [Set up an access token](#set-up-an-access-token)
2. [Configure your model training CI job](#configure-your-model-training-ci-job)

> 3. [Send and view the updates](#send-the-updates)

Reviewer: **View live metrics** is a separate H2 now.

Author: Yes. But the 3 steps are about **sending** live updates. So I didn't include **View live metrics** as a separate step here. This list is not the table of contents in this page. But let me know if you think this must be changed.

Live metrics must be sent to this ingestion end-point: TODO: Add the end-point
here

https://www.youtube.com/watch?v=hKf4twg832g

TODO: Replace this with a video tutorial for live metrics

## Set up an access token

Iterative Studio uses access tokens to authorize [DVCLive] to send live updates
to the metrics and plots.

### Generate access token

To set up the access token, open your user profile page. In the
`Studio access token` section, click on `Generate new token`.

### Regenerate access token

You can regenerate the access token at any point. The old access token will no
longer be authorized to send live metrics and plots to Iterative Studio.

### Reset (delete) access token

You can delete the access token when you no longer need it.

The option to delete the access token is also available when you try to change

> your password. That is, you can reset all your access credentials (your password

Reviewer: Reset or delete?

Author: It's called reset in one place (change password modal) and delete in another (profile page) in Studio. That is why I used both the terms in the docs. But I guess it's better to fix this in Studio itself.
@mvshmakov can we change to one term everywhere - I think delete is clear, wdyt?

Reviewer: Reset sounds like something is preserved but returned to its original condition. I think we mean delete 🙂

Author: I think so too. cc @mvshmakov

Reviewer 2: Yes, within the flow of changing the password and flow of confirming the new password after resetting it, we have the option to delete an access token.

We have a correct label in a tooltip hint that states the token will be deleted. I've also renamed the "Reset access token" to "Delete access token" in both places in the app for clarity here: https://github.com/iterative/studio/pull/4800.

So, in the docs, we need to state "delete the access token".

Author: Now, we don't have a separate section for `Reset / delete access token`. And I've reviewed the text in this page for clarity.

and the access token) at once. This is handy if you suspect that your account
security may have been compromised.

## Configure your model training CI job

> You should define the following environment variables in your CI job:

Reviewer: Feels like it needs more context, at least links, to what model training CI jobs are, what they do in Studio, how to zet them up, etc.

Author: The context is set in the first section of this page. I've revised it to make it more clear. Let me know what you think.

- `STUDIO_REPO_URL`: URL for the Git repository connected to the project. E.g.,
  `github:iterative/example-get-started`
- `STUDIO_TOKEN`: Your access token. The access token must be present in any
  request that sends data to the Iterative Studio ingestion endpoint. Requests
  with missing or incorrect access tokens are rejected with an appropriate HTTP
  error code and error message.

One way to provide the above values to your CI job is to set them as GitHub
secrets.

Here is an example GitHub action. TODO: provide link to an example snippet (such
as
[this](https://github.com/iterative/test-dvclive-studio/blob/086a51d76c7983f24c091e1b007820916aa75e7d/.github/workflows/test_live_metrics.yaml#L17-L19))

## Send the updates

> Now, whenever you log your metrics or plots using [DVCLive], they will be
> automatically sent to Iterative Studio.

Reviewer: Can it be clearer? Idea
```suggestion
When the training CI job runs or whenever you log your metrics or plots using [DVCLive], they will be
automatically sent to Iterative Studio.
```

Author: I've changed it, but slightly differently from what you have suggested. PTAL.

Iterative Studio stores the live Metrics in its database. In the project table,
the live metrics are displayed next to the Git commit corresponding to the
experiment.

## View live metrics

> The live metrics are displayed alongside the corresponding experiment commits,
> and updates to the live metrics are highlighted in orange. The running
> experiments themselves are displayed with a `Running` indicator.
> 
> TODO: Add a screenshot of the `Running` indicator
> 
> The number of experiments with new updates to the live metrics values are
> displayed in the `Live` icon, which can also be used to filter and show only
> running experiments in the table.
> 
> Live plots are displayed in the

Reviewer: Minor: terminology consistency
```suggestion
The live metrics are plotted alongside the corresponding experiment commits,
and updates to the live metrics are highlighted in orange. The running
experiments themselves are displayed with a `Running` indicator.

TODO: Add a screenshot of the `Running` indicator

The number of experiments with new updates to the live metrics values are
displayed in the `Live` icon, which can also be used to filter and show only
running experiments in the table.

Plots of the live metrics are displayed in the
```

Author: > The live metrics are **plotted** 

Not really. The metrics scalars are displayed in the project table in the experiment (commit) row. They aren't really plotted.

> Plots of the live metrics

This is also not entirely true. The live plots are updates to the **plots** files, which may or may not have the same values as the live **metrics**. @mvshmakov pls correct me if I'm wrong.

Reviewer: Oh OK. It's just unclear what "Live plots" are. I thought the history of live metrics was plotted, not that plots were updated in real time. These seem like different features but not well separated in [this page](https://dvc-org-live-metrics-bw8woadv9.herokuapp.com/doc/studio/user-guide/projects-and-experiments/live-metrics) so far. Probably adding image samples would help too.

Author: Got it. I've added `and plots` in all applicable places. Check out [this commit](https://github.com/iterative/dvc.org/pull/4012/commits/be466c550b19ef85ff638710ca849d9386fa22e3).

Reviewer 2: For clarity, we may call such experiments (as a table row) "live experiments". These experiments may contain "live metrics" and "live plots". IIRC, we may theoretically have a situation when there are no live metrics, but only live plots set up and wise-versa as well.

So, what do you think about such terminology @tapadipti?

3. [Send and view the updates](#send-the-updates)

Live metrics must be sent to this ingestion end-point: TODO: Add the end-point
here

https://www.youtube.com/watch?v=hKf4twg832g

TODO: Replace this with a video tutorial for live metrics

## Set up an access token

Iterative Studio uses access tokens to authorize [DVCLive] to send live updates
to the metrics and plots.

### Generate access token

To set up the access token, open your user profile page. In the
`Studio access token` section, click on `Generate new token`.

### Regenerate access token

You can regenerate the access token at any point. The old access token will no
longer be authorized to send live metrics and plots to Iterative Studio.

### Reset (delete) access token

You can delete the access token when you no longer need it.

The option to delete the access token is also available when you try to change
your password. That is, you can reset all your access credentials (your password
and the access token) at once. This is handy if you suspect that your account
security may have been compromised.

## Configure your model training CI job

You should define the following environment variables in your CI job:

> - `STUDIO_ENDPOINT`: The Iterative Studio ingestion endpoint. Set this to
>   `https://studio.iterative.ai/api/live`

Reviewer 3: End users don't need to set this env var, the default value already is `https://studio.iterative.ai/api/live`. The env var was added for internal usage (i.e. post to local server)

```suggestion
```

Author: Thanks @daavoo

> - `STUDIO_REPO_URL`: URL for the Git repository connected to the project. E.g.,
>   `https://github.com/iterative/example-get-started`

Reviewer 3: We can check with @Suor on the expected/supported formats but on my local tests the full URL didn't work and this was the expected format:

```suggestion
- `STUDIO_REPO_URL`: URL for the Git repository connected to the project. E.g.,
  `github:iterative/example-get-started`
```

Author: 👍

# Generate live (real-time) metrics for experiments

When you
[submit a new experiment](/doc/studio/user-guide/projects-and-experiments/run-experiments)

> from Iterative Studio, the CI/CD setup in your Git repository gets invoked. If

Author: Addressed in https://github.com/iterative/dvc.org/pull/4012/commits/22e52337f839ef931e18531b66b56eefe4b5ebfb

this setup includes a model training process, it will be triggered.

In this model training CI action, you can use [DVCLive] to send live updates to
metrics and plots back to Iterative Studio, without writing them to your Git

> repository. This requires a 3-step process:

Reviewer: Why do we want to send updates to Studio bypassing the Git repo. Shouldn't Git be the source of truth? Where will logs be stored, DVC remote? Thanks

Author: Addressed in https://github.com/iterative/dvc.org/pull/4012/commits/22e52337f839ef931e18531b66b56eefe4b5ebfb

Reviewer: > This will help to keep your Git commit history unpolluted, as you
can push only the final results of your experiments while being able to view all
intermediate results in Studio (in the form of live metrics and plots).

Still unclear in general where logs will be stored. Doesn't DVCLive use a directory in the repo normally?

Author: You can choose not to commit / push the logs to your repo. So my understanding is that when you are running experiments, you will push the live metrics and plots to Studio but won't commit / push the DVCLive output. @daavoo is this explanation correct?

Reviewer 3: Considering the MVP, I don't think there is a reason to say `This will help to keep your Git commit history unpolluted, as you can push only the final results of your experiments`. 

The MVP doesn't really help with keeping Git commit history clean and I would drop those referneces (at least until we complete the follow-up proposal about local experiments)

The MVP value is `being able to view all intermediate results in Studio` (I would add: `while the experiment is running).

Author: Updated.
